hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/associations/TokenAssociationsDumpUtils.java

```diff
@@ -0,0 +1,150 @@
+
+import com.hedera.hapi.node.state.token.TokenRelation;
+import com.hedera.node.app.bbm.DumpCheckpoint;
+import com.hedera.node.app.bbm.utils.FieldBuilder;
+import com.hedera.node.app.bbm.utils.ThingsToStrings;
+import com.hedera.node.app.bbm.utils.Writer;
+import com.hedera.node.app.service.mono.state.adapters.VirtualMapLike;
+import com.hedera.node.app.service.mono.state.virtual.entities.OnDiskTokenRel;
+import com.hedera.node.app.state.merkle.disk.OnDiskKey;
+import com.hedera.node.app.state.merkle.disk.OnDiskValue;
+import com.swirlds.base.utility.Pair;
+import com.swirlds.virtualmap.VirtualKey;
+import com.swirlds.virtualmap.VirtualMap;
+import com.swirlds.virtualmap.VirtualValue;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class TokenAssociationsDumpUtils {
+    public static void dumpModTokenRelations(
+            @NonNull final Path path,
+            @NonNull
+                    final VirtualMap<OnDiskKey<com.hedera.hapi.node.base.TokenAssociation>, OnDiskValue<TokenRelation>>
+                            associations,
+            @NonNull final DumpCheckpoint checkpoint) {
+        try (@NonNull final var writer = new Writer(path)) {
+            final var dumpableTokenRelations = gatherTokenRelations(
+                    associations, key -> TokenAssociationId.fromMod(key.getKey()), TokenAssociation::fromMod);
+            reportOnUniques(writer, dumpableTokenRelations);
+            System.out.printf(
+                    "=== mod token associations report is %d bytes at checkpoint %s%n",
+                    writer.getSize(), checkpoint.name());
+        }
+    }
+
+    public static void dumpMonoTokenRelations(
+            @NonNull final Path path,
+            @NonNull final VirtualMap<OnDiskKey<OnDiskTokenRel>, OnDiskTokenRel> associations,
+            @NonNull final DumpCheckpoint checkpoint) {
+        try (@NonNull final var writer = new Writer(path)) {
+            final var dumpableTokenRelations =
+                    gatherTokenRelations(associations, TokenAssociationId::fromMono, TokenAssociation::fromMono);
+            reportOnUniques(writer, dumpableTokenRelations);
+            System.out.printf(
+                    "=== mono token associations report is %d bytes at checkpoint %s%n",
+                    writer.getSize(), checkpoint.name());
+        }
+    }
+
+    @NonNull
+    private static <K extends VirtualKey, V extends VirtualValue>
+            Map<TokenAssociationId, TokenAssociation> gatherTokenRelations(
+                    @NonNull final VirtualMap<K, V> source,
+                    @NonNull final Function<K, TokenAssociationId> keyMapper,
+                    @NonNull final Function<V, TokenAssociation> valueMapper) {
```

Comment:
I like this approach to adapting the same iterator over the store to both mono+modular representations.